api/src/main/java/org/openmrs/api/impl/CohortServiceImpl.java

```diff
@@ -57,108 +60,146 @@ public Cohort saveCohort(Cohort cohort) throws APIException {
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getCohortsContainingPatient(org.openmrs.Patient)
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getCohortsContainingPatient(Patient patient) {
 		return dao.getCohortsContainingPatientId(patient.getPatientId());
 	}
-	
+
 	@Transactional(readOnly = true)
 	public List<Cohort> getCohortsContainingPatientId(Integer patientId) {
 		return dao.getCohortsContainingPatientId(patientId);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getCohorts(java.lang.String)
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getCohorts(String nameFragment) throws APIException {
 		return dao.getCohorts(nameFragment);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getAllCohorts()
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getAllCohorts() throws APIException {
 		return Context.getCohortService().getAllCohorts(false);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getAllCohorts(boolean)
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getAllCohorts(boolean includeVoided) throws APIException {
 		return dao.getAllCohorts(includeVoided);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getCohort(java.lang.String)
 	 */
 	@Transactional(readOnly = true)
 	public Cohort getCohort(String name) throws APIException {
 		return dao.getCohort(name);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#purgeCohort(org.openmrs.Cohort)
 	 */
 	public Cohort purgeCohort(Cohort cohort) throws APIException {
 		return dao.deleteCohort(cohort);
 	}
-	
+
+	/**
+	 * @see org.openmrs.api.CohortService#addMembershipToCohort(org.openmrs.Cohort,
+	 * 		org.openmrs.CohortMembership)
+	 */
+	public Cohort addMembershipToCohort(Cohort cohort, CohortMembership cohortMembership) throws APIException {
+        if (!cohort.contains(cohortMembership)) {
+			ValidateUtil.validate(cohortMembership);
+			cohort.addMembership(cohortMembership);
+			Context.getCohortService().saveCohort(cohort);
+		}
+		return cohort;
+	}
+
+	/**
+	 * @see org.openmrs.api.CohortService#removeMemberShipFromCohort(org.openmrs.Cohort,
+	 * 		org.openmrs.CohortMembership)
+	 */
+	public Cohort removeMemberShipFromCohort(Cohort cohort, CohortMembership cohortMembership) throws APIException {
+		if (cohort.contains(cohortMembership)) {
+			cohort.removeMembership(cohortMembership);
+			Context.getCohortService().saveCohort(cohort);
+		}
+		return cohort;
+	}
+
+	/**
+     * @see org.openmrs.api.CohortService#patientVoided(org.openmrs.Patient)
+	 */
+	public void patientVoided(Patient patient) throws APIException {
+        List<Cohort> cohorts = Context.getCohortService().getCohortsContainingPatient(patient);
+		for (Cohort cohort : cohorts) {
+			List<CohortMembership> membersWithPatient = cohort.getMembers().stream()
+					.filter(m -> m.getPatient().equals(patient))
+					.collect(Collectors.toList());
+			membersWithPatient.forEach(m -> m.setVoided(patient.getVoided()));
```

Comment:
I assume you meant cohort.removeMember(patient). I'm a bit confused. Does ending membership mean the void status has to be set? I don't understand the correlation between setting endDate and void status for a member. Currently, removeMember(patient) sets the endDate for the members for the respective patient.